### ISSUE TITLE
Problem with via-point (virtual node) when cutting an area via osmosis

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#1984]
Problematic test case that gives different answer for route Q to P when requesting "Q, P, R" vs. "Q, P" only:

```java
@Test
public void edge_splitting_bug() {
    GraphHopperOSM hopper = new GraphHopperOSM();
    hopper.setEncodingManager(EncodingManager.create("car"));
    hopper.setGraphHopperLocation(GH_LOCATION);
    hopper.setOSMFile("../core/files/lauf-200115.osm.pbf");
    hopper.setProfiles(new ProfileConfig("profile").setVehicle("car").setWeighting("fastest"));
    hopper.importOrLoad();
    GraphHopperStorage graph = hopper.getGraphHopperStorage();
    LocationIndex locationIndex = hopper.getLocationIndex();
    FlagEncoder encoder = hopper.getEncodingManager().fetchEdgeEncoders().get(0);
    final FastestWeighting weighting = new FastestWeighting(encoder);
    final BooleanEncodedValue accessEnc = weighting.getFlagEncoder().getAccessEnc();
    EdgeFilter edgeFilter = new EdgeFilter() {
        @Override
        public boolean accept(EdgeIteratorState edgeState) {
            return edgeState.get(accessEnc) && !Double.isInfinite(weighting.calcEdgeWeight(edgeState, false))
                    || edgeState.getReverse(accessEnc) && !Double.isInfinite(weighting.calcEdgeWeight(edgeState, true));
        }
    };
    GHPoint p = new GHPoint(49.529355099741494, 11.296203940756207);
    GHPoint q = new GHPoint(49.523147, 11.293001);
    GHPoint r = new GHPoint(49.53027363148555, 11.293370881574472);

    // works
    List<GHPoint> points = Arrays.asList(p, q/*, r*/);
    // does not work
//        List<GHPoint> points = Arrays.asList(p, q, r);

    List<QueryResult> queryResults = new ArrayList<>(points.size());
    for (GHPoint point : points) {
        queryResults.add(locationIndex.findClosest(point.lat, point.lon, edgeFilter));
    }
    QueryGraph queryGraph = QueryGraph.lookup(graph, queryResults);
    Dijkstra dijkstra = new Dijkstra(queryGraph, queryGraph.wrapWeighting(weighting), TraversalMode.NODE_BASED);
    Path path = dijkstra.calcPath(queryResults.get(0).getClosestNode(), queryResults.get(1).getClosestNode());
    System.out.println(path);
    assertEquals(168.369, path.getWeight(), 0.1);
    assertEquals(168369, path.getTime(), 100);
    assertEquals(1053.846, path.getDistance(), 0.1);
}
```

[lauf-200115.osm.pbf.zip](https://github.com/graphhopper/graphhopper/files/4470801/lauf-200115.osm.pbf.zip)

Here is the situation on a map:

![image](https://user-images.githubusercontent.com/129644/79140904-07b5d400-7db9-11ea-9315-ed00f753e6e4.png)

Maybe a problem due to the area cutting. When zooming into the edge near the destination it seems to be a loop:

![image](https://user-images.githubusercontent.com/129644/79141023-43e93480-7db9-11ea-94ff-d03e364ad4b8.png)

**Comment by easbar** [comment]
This map shows another strange property: Querying the location index with some point far away like `lat=0,lon=0` does not yield an error.

**Comment by easbar** [comment]
I think I ran into this or a similar issue again. In OSMReader we calculate the `towerNodeDistance` using the entire `pointList`:

https://github.com/graphhopper/graphhopper/blob/bde99a8ccd143f3852b3bc87b3bcb8b5328f8f0f/reader-osm/src/main/java/com/graphhopper/reader/osm/OSMReader.java#L675

But later we only use the inner (pillar) points to set the way geometry and we silently assume that the first/last points of `pointList` are equal to the coordinates of the base/adj nodes:

https://github.com/graphhopper/graphhopper/blob/bde99a8ccd143f3852b3bc87b3bcb8b5328f8f0f/reader-osm/src/main/java/com/graphhopper/reader/osm/OSMReader.java#L703-L704

However, with the extract I got from Osmosis this assumption was violated: The first point in the pointList did not match the base node's coordinates. Therefore in my case `towerNodeDistance = distCalc.calculateDistance(pointList) = edge.getDistance()` was different to `distCalc.calculateDistance(edge.fetchWayGeometry(FetchMode.ALL))`. This in turn lead to a situation where the sum of the distances of two virtual edges was (very) different to the 'distance' of the edge that was split by the virtual node.

I'm not sure yet if this is a data issue (osmosis corrupts the file) or a bug on our side. But in any case maybe we should add a consistency check to make sure that `pointList.get(0) == basePoint` and `pointList.get(pointList.size()-1) == adjPoint`.

**Comment by easbar** [comment]
The other thing I find strange here is that we first calculate the `towerNodeDistance` using all points, then use the result to set the edge's `distance` field, but before we store the geometry we simplify it. Should we not rather simplify the geometry first and *then* calculate the edge distance? Otherwise we store a distance that was calculated based on some pillar nodes that we do not keep which seems inconsistent.

**Comment by easbar** [comment]
Here I added a check that compares `edge.getDistance()` with `distCalc.calcDistance(edge.fetchWayGeometry(FetchMode.ALL))`: 37768ff755fdb0108205a528313061bdf77f4330.

In `GraphHopperTest` this makes four tests fail. For example in `testOneWaySubnetwork_issue1807` that uses the ESSEN map we get:

```
Suspicious distance for edge: 5 1-25 43.129 vs. 32.10738463346008, difference: 11.021615366539919
```

and in three SRTM tests that use `monaco.osm.gz` we get:

```
Suspicious distance for edge: 5 1013-259 620.436 vs. 602.6416118459674, difference: 17.794388154032617
```

**Comment by easbar** [comment]
Regarding the initial issue in the first post here (`edge_splitting_bug`): The red marker (point R) introduces a virtual node on this loop edge seen in the close-up and this means when it is included we get a different distance, because also here `edge.getDistance() != calcDistance(edge.fetchWayGeometry(FetchMode.ALL))`. With the distance check I introduced in my previous comment we also see this explicitly when importing `lauf-200115.osm.pbf`: 

```
Suspicious distance for edge: 1007 1474-643 143.957 vs. 118.18903018395105, difference: 25.767969816048947
```

**Comment by easbar** [comment]
> I'm not sure yet if this is a data issue (osmosis corrupts the file) or a bug on our side.

I think I figured this out. The problem is that after extracting a bounding box using Osmosis there can be ways that reference nodes that no longer exist in the OSM file. And this leads to wrong edge distances. See #2221

**Comment by easbar** [comment]
Fixed via #2221
